### PR TITLE
CA-337546: resource leak handling

### PR DESCRIPTION
**Pull request by edwintorok**
A simpler alternative to https://github.com/xapi-project/xen-api-libs-transitional/pull/77

### CODE REVIEW

File: resources/dune [hunk 1]
```diff
@@ -0,0 +1,4 @@
+(library
+ (name xapi_stdext_resources)
```
Comment (psafont) [inline resources/dune:2]: Can we change the name? `xapi_stdext_` is used right only in https://github.com/xapi-project/stdext, making this package be in an unexpected place. I would prefer a name like `managed_resources`, `scoped_resources`, `safe_resources` or even plain `resources`.

File: resources/test/scoped_dropable_test.ml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  Alcotest.(check int "active counters" 0 !active) ;
+  Alcotest.check_raises "should raise" (Failure "TEST") (fun () ->
+      Scope.local
+      @@ fun (module L) ->
+      let _ = SafeOuter.create true L.scope in
+      ()) ;
+  Alcotest.(check int "active counters end" 0 !active)
+
+let tests =
+  [(__LOC__, `Quick, test_nesting); (__LOC__, `Quick, test_nesting_raised)]
```
Comment (psafont) [inline resources/test/scoped_dropable_test.ml:93]: Why can't these tests be enabled?

**Comment by psafont** [comment]
travis is failing because the new opam package is not being pinned in `.travis.yml`

File: resources/dune (hunk 1 above)
Comment (edwintorok) [reply resources/dune]: I renamed it to Safe_resources. Originally my plan was to move this to stdext, but that requires tagging a new xs-opam, which adds additional complexity in merging.

File: resources/test/scoped_dropable_test.ml (hunk 2 above)
Comment (edwintorok) [reply resources/test/scoped_dropable_test.ml]: They were for the Scope module that I deleted to simplify the code. I've made a change to drop the unused test now.

**Comment by edwintorok** [comment]
I've fixed some behavioural differences in the fixups: I was mistakenly retrying and looping if the [f] function in with_reusable_stunnel failed for whatever reason, but we only need to retry if the connection fails.

File: resources/table.ml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  let copy (t, m) =
+    Mutex.execute m (fun () -> Hashtbl.copy t), Mutex.create ()
+
+  let length (t, m) =
+    Mutex.execute m (fun () -> Hashtbl.length t)
+
+  let move_into (t, m) k v =
+    let v = V.move_out_exn v in
+    Mutex.execute m (fun () -> Hashtbl.add t k v)
```
Comment (lindig) [inline resources/table.ml:30]: `Hashtbl.add` would shadow an existing entry. Is this correct or would it be safer to use `Hashtbl.replace`?

File: resources/unixfd.ml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+   OCaml already provides for channels, see [Sys.enable_runtime_warnings]. *)
+let on_finalise_leaked (chan, loc) =
+  let enabled = Sys.runtime_warnings_enabled () in
+  if enabled then
+    Printf.eprintf "[unix_fd]: resource leak detected, allocated at %s\n%!" loc ;
+  try Unix.close chan
+  with e ->
+    if enabled then (
+      Printexc.print_backtrace stderr ;
+      Printf.eprintf "[unix_fd]: close failed: %s\n%!" (Printexc.to_string e)
```
Comment (lindig) [inline resources/unixfd.ml:32]: Should this also report `loc`?

Comment (edwintorok) [reply resources/unixfd.ml]: Updated

File: resources/table.ml (hunk 3 above)
Comment (edwintorok) [reply resources/table.ml]: It seems the old code used add, but I agree about replace. Also have to release the old value, so it is a few more lines, see my latest commit.

**Comment by edwintorok** [comment]
Travis passed: https://travis-ci.org/github/xapi-project/xen-api-libs-transitional/builds/692652186
Once https://github.com/xapi-project/xen-api/pull/4097 is approved we can merge both.

**Pull request merged**